content/precise-types.md

```diff
@@ -0,0 +1,517 @@
+  val bb1 = idBoxBox(id(1)) //: Box[1] 
+  ~~~
+  * It is an Alias of a precise type parameter.
+  ~~~ scala
+  class PreciseBox[precise T]
+  //`A` is precise because it is aliasing the precise type parameter `T`
+  type PB[A] = PreciseBox[A]
+  ~~~
+  * It is introduced by an applied type and it is in a precise position of that type. 
+    A precise position can present directly, when the position parameter is annotated with `precise`, or indirectly by "boring" through an applied type composition so that all applied type parameters are precise if the applied type itself is in a precise position.
+  ~~~ scala
+  class PreciseBox[precise T]
+  //`BB` is precise because it was introduced by `PreciseBox` in the precise position `T`
+  def pbox[BB](x: PreciseBox[BB]): Box[BB] = ??? 
+  //`BB` is imprecise because it was not introduced by `PreciseBox` and is not annotated with precise
+  def pbox[BB](b: BB)(x: PreciseBox[BB]): Box[BB] = ??? 
+  class Cov[+T]
+  //`BB` is precise because it was introduced by `Cov` which is applied in `PreciseBox` in the precise position `T`
+  def pboxcov[BB](x: PreciseBox[Cov[BB]]): Box[BB] = ??? 
+  ~~~
+  * It is a upper-bounded by precise type.
+    Upper-bounds may manifest either directly in a type param declaration or dynamically constrained by applying one method as an argument of another.
+  ~~~ scala
+  //`T` is precise because it is upper-bounded by `P`
+  def id[precise P, T <: P](t: T): Box[T] = ??? 
+  ~~~
+  * It is a by-name reference with a precise result type
+  ~~~ scala
+  //`: => T` is precisely typed because the result `T` is precise
+  def id[precise T](t: => T): Box[T] = ??? 
+  ~~~
+  * It is a FuncXX type with a precise result type
+  ~~~ scala
+  //`: (Int, Int) => T` is precisely typed because the result `T` is precise
+  def id[precise T](t: (Int, Int) => T): Box[T] = ??? 
+  ~~~
+
+#### Precise Term Arguments
+A term expression is *precisely typed* when it is applied in a term argument position that is annotated by a precise type. 
+~~~ scala
+//`t` argument is precise because it is annotated by the precise param `T` 
+def id[precise T](t: T): Box[T] = ??? 
+~~~
+**Tuple arguments special case**
+We also support a special case where tuple of terms applied on an argument annotated with tuple type of the same arity. This causes each part of the tuple term to be precisely typed or not according to the specific precise type of that position.
+~~~ scala
+def id[precise T1, T2](t: (T1, T2)): Box[(T1, T2)] = ??? 
+//`(1, 2)` is precisely typed because `T1` is precise
+//`3` is not precisely typed because `T2` is not precise
+val x = id(((1, 2), 3)) //: Box[((1, 2), Int)]
+~~~
+
+
+#### Precise Typing (of Term Argument Expressions)
+Precise typing (mode) is activated when an expression is applied in a precise argument position. Here are the same examples given in the motivation, but now precisely typed due to the `precise` modifier application:
```

Comment:
I'll begin with properly motivating the reasoning behind the precise typing mode. Singletons are non-issue, but when we have expression compositions then we need to consider whether or not we want to support precise typing of such expressions. Tuples are a good example because we cannot change their definition due to backward compatibility (add precise annotation on their parameters), but we still need precise tuples, so what can we do? We can just special-case tuples, but from the discussion thread on contributors I understand that's undesired. So we end up with the only option (I could think about) - an ability to define any applied expression as precise. That is where precise typing (mode) comes in.

Reply:
> * The cause for the difference in typing is invisible. There could be some expected type anywhere further out that is propagated by the proposed rules of propagation from a far away place that changes how types are inferred. How to troubleshoot that?

The mode is not that far reaching. It's only active within an expression composition (and some minor exceptions).